File: targets/omq_rb_target/target.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# zmq-arena target wrapper for omq.rb, the pure-Ruby ZMTP implementation.
#
# Not a binding: omq.rb speaks the wire protocol itself, with no libzmq, no FFI
# and no C extension. It is measured as an implementation rather than as a
# language access path, which is why `impl` is native below.
#
# The wrapper times its own steady-state window and prints one result line. A
# Ruby process spends a hundred-odd milliseconds booting and requiring before it
# reaches the benchmark, and letting the harness time the process would charge
# that to the measurement.

require 'json'
require 'async'

VARIANT = (i = ARGV.index('--variant')) ? ARGV[i + 1] : 'default'

if ARGV[0] == 'describe'
  require 'omq'
  # The engine is omq.rb itself: it implements ZMTP in Ruby rather than binding
  # anything. Grouping it with the omq family is the dashboard's job, through
  # the registry's `family` field, and does not belong in a target's report of
  # what it actually is.
  puts JSON.generate(
    engine: 'omq.rb',
    lib_version: OMQ::VERSION,
    binding_version: nil,
    lib_language: 'Ruby',
    impl: 'native',
    ffi_to: nil,
    language: 'Ruby',
    concurrency: 'async',
    threading: 'single',
    io: 'epoll'
  )
  exit 0
end

require 'omq'

def arg(name, default = nil)
  i = ARGV.index("--#{name}")
  i ? ARGV[i + 1] : default
end

def flag?(name) = ARGV.include?("--#{name}")

def knob(key)
  ARGV.each_with_index do |a, i|
    next unless a == '--knob'
    k, _, v = ARGV[i + 1].to_s.partition('=')
    return v if k == key
  end
  ENV["ARENA_KNOB_#{key.upcase}"]
end

ROLE     = arg('role')
KIND     = arg('kind', 'throughput')
ENDPOINT = arg('endpoint')
PAYLOAD  = 'x' * arg('payload-bytes').to_i
MESSAGES = arg('messages').to_i
WARMUP   = arg('warmup', '0').to_i
DURATION = arg('duration-secs', '0').to_f
BIND     = flag?('bind')

warn "omq-rb-target: role=#{ROLE} kind=#{KIND} ep=#{ENDPOINT} " \
     "payload=#{PAYLOAD.bytesize}B msgs=#{MESSAGES} variant=#{VARIANT}"

# The arena forwards one knob superset to every target; apply what this engine
# understands and ignore the rest, per the knob convention.
def tune(sock)
  if (v = knob('sndhwm')) && sock.respond_to?(:send_hwm=)
    sock.send_hwm = v.to_i
  end
  if (v = knob('rcvhwm')) && sock.respond_to?(:recv_hwm=)
    sock.recv_hwm = v.to_i
  end
  sock
end

def open_socket(type, endpoint, bind)
  sock = bind ? type.bind(endpoint) : type.connect(endpoint)
  tune(sock)
end

def print_latency(rtts)
  if rtts.empty?
    puts 'LATENCY 0 0 0 0 0 0 0'
    return
  end
  rtts.sort!
  q = ->(p) { rtts[[(rtts.size * p).to_i, rtts.size - 1].min] }
  puts format('LATENCY %d %d %d %d %d %d %d',
              rtts.size, rtts.first, q[0.50], q[0.90], q[0.99], q[0.999], rtts.last)
end

# Wait a bounded time for the first message, matching the compiled targets:
# they give up after ten seconds and report a zero window rather than blocking.
# Without this a starved cell would be a missing data point here and a zero
# everywhere else, which is not a comparison.
def first_or_bail(sock, seconds = 10)
  sock.read_timeout = seconds if sock.respond_to?(:read_timeout=)
  sock.receive
  sock.read_timeout = nil if sock.respond_to?(:read_timeout=)
  true
rescue StandardError
  puts 'THROUGHPUT 0 0.000001'
  false
end

def timed_drain(sock, seconds)
  return unless first_or_bail(sock)
  count = 1
  t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  deadline = t0 + seconds
  while Process.clock_gettime(Process::CLOCK_MONOTONIC) < deadline
    sock.receive
    count += 1
  end
  secs = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0
  puts format('THROUGHPUT %d %.6f', count, [secs, 1e-9].max)
end

# Everything runs inside an Async reactor.
#
# omq.rb is fibre-based and async-native. Outside a reactor it still works, on a
# shared IO thread, and its README offers that so simple scripts do not need
# boilerplate -- but it is not the path the library is built for. Measured on
# this wrapper, one process, 64B over TCP: 25k msgs/s on the shared IO thread
# against 628k inside a reactor, a factor of twenty-five. Benchmarking the
# convenience path would have published omq.rb as forty times slower than a
# Python binding, which says nothing about the library and everything about how
# the harness drove it.
Async do |task|
case KIND
when 'throughput'
  if ROLE == 'sub'
    pull = open_socket(OMQ::PULL, ENDPOINT, true)
    WARMUP.times { pull.receive }   # drained before the clock starts
    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    MESSAGES.times { pull.receive }
    secs = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0
    puts format('THROUGHPUT %d %.6f', MESSAGES, [secs, 1e-9].max)
  else
    push = open_socket(OMQ::PUSH, ENDPOINT, false)
    (MESSAGES + WARMUP).times { push << PAYLOAD }
    push.close # drains queued frames rather than dropping them
  end

when 'latency'
  if ROLE == 'sub'
    rep = open_socket(OMQ::REP, ENDPOINT, true)
    loop do
      msg = rep.receive
      rep << msg.first
    end
  else
    req = open_socket(OMQ::REQ, ENDPOINT, false)
    WARMUP.times { req << PAYLOAD; req.receive }
    rtts = []
    MESSAGES.times do
      t = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
      req << PAYLOAD
      req.receive
      rtts << (Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - t)
    end
    print_latency(rtts)
  end

when 'pubsub'
  if ROLE == 'pub'
    pub = open_socket(OMQ::PUB, ENDPOINT, true)
    # PUB drops what it publishes before a subscriber has finished subscribing.
    # Worse here than elsewhere: omq.rb is fibre-based, and a tight publish loop
    # never reaches an await point, so a subscriber arriving after the loop
    # starts can never complete its handshake and receives nothing at all. That
    # is why this cell reported zero before. The settle has to cover the arrival
    # of every subscriber, not just the first.
    sleep(ENV.fetch('ARENA_PUB_SETTLE', '2.0').to_f)
    n = 0
    loop do
      pub << PAYLOAD
      n += 1
      # Hand the reactor back regularly so it can service subscribers and
      # actually write the socket, rather than spinning in this fibre.
      task.yield if (n % 256).zero?
    end
  else
    sub = open_socket(OMQ::SUB, ENDPOINT, false)
    sub.subscribe('')
    timed_drain(sub, DURATION)
  end

when 'fanout', 'fanin'
  if ROLE == 'pub'
    push = open_socket(OMQ::PUSH, ENDPOINT, BIND)
    loop { push << PAYLOAD }
  else
    pull = open_socket(OMQ::PULL, ENDPOINT, BIND)
    timed_drain(pull, DURATION)
  end

else
  warn "omq-rb-target: unsupported kind #{KIND}"
  exit 1
end
# `Async { }` reports a failed task to the logger and returns normally, so a
# publisher that dies mid-window still exits 0 and reads as a healthy feeder.
# For a pubsub or fanout cell that is the worst outcome available: the consumer
# on the other side keeps timing a stream nobody is feeding any more and reports
# real arithmetic over a starved window. `wait` re-raises, which makes the
# process exit non-zero and the cell fail instead of publishing the number.
end.wait

File: targets/jeromq_target/src/Target.java
// zmq-arena target wrapper for JeroMQ, the pure Java ZMTP implementation.
//
// Not a binding: JeroMQ reimplements the protocol on the JVM, with no libzmq
// and no JNI, so it is measured as an implementation.
//
// The wrapper times its own steady-state window, which matters more here than
// for any other target so far. A JVM spends hundreds of milliseconds starting
// and runs interpreted until the JIT promotes the hot paths; letting the
// harness time the process would charge all of that to the benchmark and report
// it as the language being slow at messaging. The matrix's warmup messages are
// what get those paths compiled before the clock starts.
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ;

public final class Target {

    static String arg(String[] a, String name, String dflt) {
        int i = Arrays.asList(a).indexOf("--" + name);
        return (i >= 0 && i + 1 < a.length) ? a[i + 1] : dflt;
    }

    static boolean flag(String[] a, String name) {
        return Arrays.asList(a).contains("--" + name);
    }

    static String knob(String[] a, String key) {
        for (int i = 0; i < a.length - 1; i++) {
            if (!a[i].equals("--knob")) continue;
            String[] kv = a[i + 1].split("=", 2);
            if (kv.length == 2 && kv[0].equals(key)) return kv[1];
        }
        return System.getenv("ARENA_KNOB_" + key.toUpperCase());
    }

    static void describe() {
        // JeroMQ reports the libzmq protocol version it implements; its own
        // version is the artifact version baked in at build time.
        String ver = System.getProperty("arena.jeromq.version", "unknown");
        System.out.println("{\"engine\":\"jeromq\",\"lib_version\":\"" + ver + "\","
            + "\"binding_version\":null,\"lib_language\":\"Java\",\"impl\":\"native\","
            + "\"ffi_to\":null,\"language\":\"Java\",\"concurrency\":\"sync\","
            + "\"threading\":\"native\",\"io\":\"epoll\"}");
    }

    // The arena forwards one knob superset to every target; apply what this
    // engine understands and ignore the rest.
    static void tune(ZMQ.Socket s, String[] a) {
        String v;
        if ((v = knob(a, "sndhwm")) != null) s.setSndHWM(Integer.parseInt(v));
        if ((v = knob(a, "rcvhwm")) != null) s.setRcvHWM(Integer.parseInt(v));
    }

    static void printLatency(List<Long> rtts) {
        if (rtts.isEmpty()) {
            System.out.println("LATENCY 0 0 0 0 0 0 0");
            return;
        }
        Collections.sort(rtts);
        int n = rtts.size();
        System.out.println("LATENCY " + n + " " + rtts.get(0) + " "
            + rtts.get(Math.min((int) (n * 0.50), n - 1)) + " "
            + rtts.get(Math.min((int) (n * 0.90), n - 1)) + " "
            + rtts.get(Math.min((int) (n * 0.99), n - 1)) + " "
            + rtts.get(Math.min((int) (n * 0.999), n - 1)) + " "
            + rtts.get(n - 1));
    }

    static void timedDrain(ZMQ.Socket s, double seconds) {
        s.recv(0); // first message: the link is live, clock starts after it
        long count = 1;
        long t0 = System.nanoTime();
        long deadline = t0 + (long) (seconds * 1e9);
        while (System.nanoTime() < deadline) {
            s.recv(0);
            count++;
        }
        double secs = Math.max((System.nanoTime() - t0) / 1e9, 1e-9);
        System.out.printf("THROUGHPUT %d %.6f%n", count, secs);
    }

    public static void main(String[] args) {
        if (args.length > 0 && args[0].equals("describe")) {
            describe();
            return;
        }
        String role = arg(args, "role", "sub");
        String kind = arg(args, "kind", "throughput");
        String endpoint = arg(args, "endpoint", null);
        int payloadBytes = Integer.parseInt(arg(args, "payload-bytes", "0"));
        long messages = Long.parseLong(arg(args, "messages", "0"));
        long warmup = Long.parseLong(arg(args, "warmup", "0"));
        double duration = Double.parseDouble(arg(args, "duration-secs", "0"));
        boolean bind = flag(args, "bind");
        byte[] payload = new byte[payloadBytes];
        Arrays.fill(payload, (byte) 'x');

        System.err.println("jeromq-target: role=" + role + " kind=" + kind
            + " ep=" + endpoint + " payload=" + payloadBytes + "B msgs=" + messages);

        // One IO thread unless the matrix says otherwise: the arena gives every
        // engine the same lane budget rather than letting each size its own.
        String io = knob(args, "io_threads");
        try (ZContext ctx = new ZContext(io == null ? 1 : Integer.parseInt(io))) {
            switch (kind) {
                case "throughput": {
                    if (role.equals("sub")) {
                        ZMQ.Socket pull = ctx.createSocket(SocketType.PULL);
                        tune(pull, args);
                        pull.bind(endpoint);
                        for (long i = 0; i < warmup; i++) pull.recv(0);
                        long t0 = System.nanoTime();
                        for (long i = 0; i < messages; i++) pull.recv(0);
                        double secs = Math.max((System.nanoTime() - t0) / 1e9, 1e-9);
                        System.out.printf("THROUGHPUT %d %.6f%n", messages, secs);
                    } else {
                        ZMQ.Socket push = ctx.createSocket(SocketType.PUSH);
                        tune(push, args);
                        push.connect(endpoint);
                        for (long i = 0; i < messages + warmup; i++) push.send(payload, 0);
                        // Drain queued frames rather than dropping them, which
                        // the no-data-dropping rule forbids.
                        push.setLinger(5000);
                    }
                    break;
                }
                case "latency": {
                    if (role.equals("sub")) {
                        ZMQ.Socket rep = ctx.createSocket(SocketType.REP);
                        rep.bind(endpoint);
                        while (!Thread.currentThread().isInterrupted()) {
                            byte[] m = rep.recv(0);
                            if (m == null) break;
                            rep.send(m, 0);
                        }
                    } else {
                        ZMQ.Socket req = ctx.createSocket(SocketType.REQ);
                        req.connect(endpoint);
                        for (long i = 0; i < warmup; i++) {
                            req.send(payload, 0);
                            req.recv(0);
                        }
                        List<Long> rtts = new ArrayList<>((int) messages);
                        for (long i = 0; i < messages; i++) {
                            long t = System.nanoTime();
                            req.send(payload, 0);
                            req.recv(0);
                            rtts.add(System.nanoTime() - t);
                        }
                        printLatency(rtts);
                    }
                    break;
                }
                case "pubsub": {
                    if (role.equals("pub")) {
                        ZMQ.Socket pub = ctx.createSocket(SocketType.PUB);
                        tune(pub, args);
                        pub.bind(endpoint);
                        // PUB drops what it sends before a subscriber has
                        // finished subscribing, so settle before flooding.
                        Thread.sleep(500);
                        while (true) pub.send(payload, 0);
                    } else {
                        ZMQ.Socket sub = ctx.createSocket(SocketType.SUB);
                        tune(sub, args);
                        sub.connect(endpoint);
                        sub.subscribe(new byte[0]);
                        timedDrain(sub, duration);
                    }
                    break;
                }
                case "fanout":
                case "fanin": {
                    if (role.equals("pub")) {
                        ZMQ.Socket push = ctx.createSocket(SocketType.PUSH);
                        tune(push, args);
                        if (bind) push.bind(endpoint); else push.connect(endpoint);
                        while (true) push.send(payload, 0);
                    } else {
                        ZMQ.Socket pull = ctx.createSocket(SocketType.PULL);
                        tune(pull, args);
                        if (bind) pull.bind(endpoint); else pull.connect(endpoint);
                        timedDrain(pull, duration);
                    }
                    break;
                }
                default:
                    System.err.println("jeromq-target: unsupported kind " + kind);
                    System.exit(1);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        System.out.flush();
    }
}
